update to ocean regridding methodology

mariahpope opened the pull request:
Getting this pull request started!
These are the major changes that need to happen. There are a few other things we should probably add in, like the yaml specifying regridding type instead of that happening within the code, etc.

mariahpope commented:
@timothyas this is a first pass at adding in the new methodology!

timothyas commented on ufs2arco/regrid/mom6regridder.py:
```diff
@@ -62,16 +63,58 @@ def _create_regridder(self, ds_in: xr.Dataset) -> None:
+        # create input dataset
+        mom6_grid = self.config.get("rotation_file", None)
+        mom6_grid = xr.open_dataset(mom6_grid)
+        grid_in = xr.Dataset()
+        grid_in["lon"] = mom6_grid["geolon"]
+        grid_in["lat"] = mom6_grid["geolat"]
+        grid_in["lon_u"] = mom6_grid["geolon_u"]
+        grid_in["lat_u"] = mom6_grid["geolat_u"]
+        grid_in["lon_v"] = mom6_grid["geolon_v"]
+        grid_in["lat_v"] = mom6_grid["geolat_v"]
+        ny, nx = grid_in["lon"].shape
+        lon_b = np.empty((ny + 1, nx + 1))
+        lat_b = np.empty((ny + 1, nx + 1))
+        lon_b[1:, 1:] = mom6_grid["geolon_c"].values
+        lat_b[1:, 1:] = mom6_grid["geolat_c"].values
+        # periodicity
+        lon_b[:, 0] = lon_b[:, -1]
+        lat_b[:, 0] = lat_b[:, -1]
+        # south edge
+        dy = (lat_b[2, :] - lat_b[1, :]).mean()
+        lat_b[0, 1:] = lat_b[1, 1:] - dy
+        lon_b[0, 1:] = lon_b[1, 1:]
+        # corner point
+        lon_b[0, 0] = lon_b[1, 0]
+        lat_b[0, 0] = lat_b[0, 1]
+        grid_in["lon_b"] = xr.DataArray(data=lon_b)
+        grid_in["lat_b"] = xr.DataArray(data=lat_b)
```
Soft suggestion: we could put this in a helper method/function to declutter. Up to you though

mariahpope replied:
yeah good call. was thinking this myself but just didn't get to it. can definitely do that

timothyas reviewed (changes requested):
Looks good @mariahpope! Thanks. I just have a couple small suggestions as it stands right now. Let me know how you'd like to handle the additional edits, whether you'd like to add them to this PR or make a separate PR.

mariahpope commented on ufs2arco/regrid/mom6regridder.py:
```diff
@@ -3,6 +3,7 @@
 import warnings
 import xarray as xr
 import numpy as np
+import cf_xarray as cfxr
```
done!

mariahpope commented on ufs2arco/regrid/mom6regridder.py:
```diff
@@ -62,16 +63,58 @@ def _create_regridder(self, ds_in: xr.Dataset) -> None:
             )
 
+        # create input dataset
+        mom6_grid = self.config.get("rotation_file", None)
+        mom6_grid = xr.open_dataset(mom6_grid)
```
done!

mariahpope commented on docs/config-replay.yaml:
```diff
@@ -102,7 +102,7 @@ MOM6Dataset:
 
     regrid:
-        rotation_file:
+        rotation_file: mom6_grid_0.25_degree.nc
```
done.

mariahpope commented:
@timothyas thanks for the review. made the some updates and definitely decluttered a bit. let me know if you think of anything else! trying to think through anything we may be forgetting that will be needed for this to actually run. its all working great on little local tests though so that is encouraging!

timothyas commented on ufs2arco/regrid/ufsregridder.py:
```diff
@@ -182,8 +182,12 @@ def regrid_tripolar(
                 # rotate to earth-relative
-                urot = interp_u * ds_rot.cos_rot + interp_v * ds_rot.sin_rot
-                vrot = interp_v * ds_rot.cos_rot - interp_u * ds_rot.sin_rot
+                urot = (
+                    interp_u * ds_rot.cos_rot.values + interp_v * ds_rot.sin_rot.values
+                )
+                vrot = (
+                    interp_v * ds_rot.cos_rot.values - interp_u * ds_rot.sin_rot.values
+                )
```
Out of curiosity, is there a reason for adding the `.values` accessor here?

mariahpope replied:
I think it just made it go faster... for some reason it was taking a sec on my end.

mariahpope replied:
I actually went and refactored cos/sin a little in general to make it neater so this went back to the old way (without .values)

timothyas replied:
Ah that makes sense. I guess they were dask arrays originally. Thanks for looking into that

timothyas commented:
Thanks @mariahpope for humoring me on the decluttering comment :) I think this looks good, I also can't think of any necessary additions at the moment. Just one more question from me and I think we can merge this.

timothyas commented:
#17 is merged so feel free to drop in the regridding statements, also feel free to turn off draft mode once you do since this PR is basically ready

mariahpope commented:
@timothyas alright I added all regridding statements. Did some local testing and found a few bugs I fixed as well. At this point I was able to fully run the pipeline for mom6 locally on a number of time steps and the output looked as expected!

timothyas commented on examples/replay/mom6_grid_0.25_degree.nc:
Can you remove this from the repo? We already have it publicly available elsewhere and it's a lot to carry around (not recommended by github).

mariahpope replied:
good call. done.

timothyas commented on examples/replay/mom6_grid_1.00_degree.nc:
Same here, can you remove this?

mariahpope replied:
done.

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,78 @@ FV3Dataset:
+
+    cycles:
+        start: "1994-01-01T00"
+        end: "1999-06-13T06"
```
```suggestion
        end: "2023-12-31T06"
```

mariahpope replied:
done.

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,78 @@ FV3Dataset:
+        freq: "6h"
+
+    time:
+        start: "1993-12-31T18"
+        end: "1999-06-13T03"
```
The cycles and time will look a little different in the ocean, since it doesn't have the IAU, which modifies the forecast hour time with a -6 hour offset from the cycle. I'm happy to talk more about that if you want. It's weird.
```suggestion
        start: "1994-01-01T00"
        end: "2023-12-31T06"
```

mariahpope replied:
done.

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,78 @@ FV3Dataset:
+
+    cycles:
+        start: "1994-01-01T00"
+        end: "1999-06-13T06"
```
```suggestion
        end: "1999-06-14T00"
```

mariahpope replied:
done.

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,78 @@ FV3Dataset:
+        freq: "6h"
+
+    time:
+        start: "1993-12-31T18"
+        end: "1999-06-13T03"
```
```suggestion
        start: "1994-01-01T00"
        end: "1999-06-14T00"
```

mariahpope replied:
done.

timothyas commented on examples/replay/replay_mover.py:
```diff
@@ -162,13 +162,42 @@ def run(self, job_id):
+            )
+        lons = gaussian_grid['grid_xt'].values
+        lats = gaussian_grid['grid_yt'].values
+        rg = MOM6Regridder(
```
```suggestion
        rg = self.Regridder(
```

timothyas replied:
And could you modify the __init__ method with something like this:

```python
        self.Regridder = None
        if component.lower() == "fv3":
            self.Dataset = FV3Dataset
            self.cached_path = self.fv3_path
            if "regrid" in self.config.keys():
                raise NotImplementedError
        elif component.lower() == "mom6":
            self.Dataset = MOM6Dataset
            self.cached_path = self.mom6_path
            self.Regridder = MOM6Regridder
        elif component.lower() == "cice6":
            self.Dataset = CICE6Dataset
            self.cached_path = self.cice6_path
            self.Regridder = CICE6Regridder

mariahpope replied:
done.

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,78 @@ FV3Dataset:
+        weights_file_t2t:
+        weights_file_u2t:
+        weights_file_v2t:
+        periodic: True
```
Could you also move the "interp_method" option to the config file? I think it would be a nice feature for reproducibility. Specifically I'm thinking:
- remove the option from UFSRegridder initialization (and all child classes)
- instead set the object's attribute from the config file

Actually it shouldn't be too bad.

mariahpope replied:
done.

timothyas reviewed (changes requested):
Very nice @mariahpope! This looks really clean, and I'm glad it's giving expected output.

I have just a couple more detailed comments, e.g. I realized we'll have to change the time bounds for the ocean, and also a couple of implementation requests. Please let me know if you agree/disagree with any of that.

mariahpope commented:
@timothyas getting close! and good call on the interp method. I think all comments were addressed. So we should be almost good to go!

I do have one question... we currently have this as the information for mom6 dates in the yaml. the start and end make sense to me... but I wanted to make sure that the forecast hours and freq that i updated to are correct!
    forecast_hours: [0]
    cycles:
        start: "1994-01-01T00"
        end: "1999-06-14T00"
        freq: "6h"
    time:
        start: "1994-01-01T00"
        end: "1999-06-14T00"
        freq: "6h"

timothyas commented:
@mariahpope this looks good to me! Thanks for addressing all of that. I think the dates you mentioned in the comment are correct. I didn't realize it when I moved the atmosphere data, but it looks like we have ~1 more day of data so may as well grab it (i.e. out to june 14 00h). And I think the frequency (6h) and forecast hours [0] you have are correct. I'll ping Phil to make sure he agrees.

mariahpope commented:
> @mariahpope this looks good to me! Thanks for addressing all of that. I think the dates you mentioned in the comment are correct. I didn't realize it when I moved the atmosphere data, but it looks like we have ~1 more day of data so may as well grab it (i.e. out to june 14 00h). And I think the frequency (6h) and forecast hours [0] you have are correct. I'll ping Phil to make sure he agrees.

sounds good! once we have that confirmed I think we should be good to go!

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,79 @@ FV3Dataset:
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/03h-freq/zarr/"
+    coords_path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/coordinates/zarr/"
+    forecast_hours: [0]
```
I was wrong, I'm glad we checked with Phil. The following would be consistent with the atmosphere data that's already online
```suggestion
    forecast_hours: [-6]
```

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,79 @@ FV3Dataset:
+        freq: "6h"
+
+    time:
+        start: "1994-01-01T00"
+        end: "2023-12-31T06"
```
These are offset from the cycles by -6hr. 
```suggestion
        start: "1993-12-31T18"
        end: "2023-12-31T00"
```

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,79 @@ FV3Dataset:
+    time:
+        start: "1994-01-01T00"
+        end: "2023-12-31T06"
+        freq: "3h"
```
```suggestion
        freq: "6h"
```

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,79 @@ FV3Dataset:
+    file_prefixes:
+        - ocn_
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/03h-freq/zarr/"
```
I just realized we should also change this outer directory to reflect the 06h frequency
```suggestion
    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/06h-freq/zarr/"
```

timothyas commented on examples/replay/config-0.25-degree.yaml:
```diff
@@ -211,3 +211,79 @@ FV3Dataset:
+        - ocn_
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/03h-freq/zarr/"
+    coords_path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/0.25-degree/coordinates/zarr/"
```
This coordinates stuff is not really necessary actually
```suggestion
```

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,79 @@ FV3Dataset:
+    file_prefixes:
+        - ocn_
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/03h-freq/zarr/"
```
```suggestion
    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/06h-freq/zarr/"
```

mariahpope replied:
updated.

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,79 @@ FV3Dataset:
+        - ocn_
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/03h-freq/zarr/"
+    coords_path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/coordinates/zarr/"
```
```suggestion
```

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,79 @@ FV3Dataset:
+    
+    path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/03h-freq/zarr/"
+    coords_path_out: "gcs://noaa-ufs-gefsv13replay/ufs-hr1/1.00-degree/coordinates/zarr/"
+    forecast_hours: [0]
```
```suggestion
    forecast_hours: [-6]
```

timothyas commented on examples/replay/config-1.00-degree.yaml:
```diff
@@ -218,3 +218,79 @@ FV3Dataset:
+        freq: "6h"
+
+    time:
+        start: "1994-01-01T00"
+        end: "1999-06-14T00"
```
```suggestion
        start: "1993-12-31T18"
        end: "1999-06-13T18"
```

timothyas reviewed (changes requested):
Hey @mariahpope I dropped in some suggested changes following Phil's email. Please let me know if this looks inconsistent. Once these changes are done I think we're ready to merge, and ready to start moving the datasets! 🚀

mariahpope commented:
@timothyas alright made changes to yaml!

timothyas reviewed (approved):
Looks good @mariahpope! I will go ahead and merge. Thanks for all your work on this
